src/ivc/protogalaxy/mod.rs

```diff
@@ -183,8 +183,53 @@ mod verify_chip {
         }
     }
 
+    /// Powers of one assigned value counted on-circuit
+    ///
+    /// Since powers are required many times during synthesis, let's keep these values separate
+    pub struct ValuePowers<F: PrimeField> {
+        powers: Vec<AssignedValue<F>>,
+    }
+
+    impl<F: PrimeField> ValuePowers<F> {
+        pub fn new(one: AssignedValue<F>, value: AssignedValue<F>) -> Self {
+            Self {
+                powers: vec![one, value],
+            }
+        }
+
+        pub fn iter(&self) -> impl Iterator<Item = &AssignedValue<F>> {
+            self.powers.iter()
+        }
+
+        pub fn value(&self) -> &AssignedValue<F> {
+            self.powers
+                .get(1)
+                .expect("Cannot be created without at least one element inside")
+        }
+
+        pub fn get_or_eval<const T: usize>(
+            &mut self,
+            region: &mut RegionCtx<F>,
+            main_gate: &MainGate<F, T>,
+            index: usize,
+        ) -> Result<AssignedValue<F>, Halo2PlonkError> {
+            if let Some(value) = self.powers.get(index) {
+                return Ok(value.clone());
+            }
+
+            while self.powers.len() <= index {
+                let value = self.value();
+                let last = self.powers.last().unwrap();
```

Comment:
should each time double the last value instead of multiply by first element? i.e.

```
{delta, delta^2, delta^4,...}
```

Reply:
Is not related to `delta`, this is just a cache for `x^0, x^1, x^2,...`.

Doc updated for this method\struct